### PR TITLE
`Lecture`: Fix embedding of TUM-Live videos in lecture units

### PR DESCRIPTION
**Pull request by JohannesWt**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).

### Motivation and Context

This PR resolves an existing issue where a video unit, added to a lecture using a TUM-Live URL without the 'video_only=1' URL parameter, could not be embedded. [#8157](https://github.com/ls1intum/Artemis/issues/8157)

### Description

- A TUM Live URL added to the source field of a video unit without the search param 'video_only=1' will not be accepted
- A TUM Live URL without the search param 'video_only' can be transformed using the 'transform into embeddable format' button to a valid TUM Live URL

### Steps for Testing

Prerequisites:
- 1 Lecture to add (or edit) a video unit

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to Lectures
4. Click on "Units" on a lecture
5. Click on "Video" in the button bar to add a new video
6. Insert a name for the video unit
7. Insert a TUM live link (e.g. https://live.rbg.tum.de/w/AdvProg/35931) into the field "Transform into embeddable format"
8. Click the arrow button next to the field
9. Click on "Submit"
10. Expand the created video unit on the video unit overview page
11. Play the video

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
	- Enhanced video URL validation and transformation for lecture units, including support for TUM-Live URLs.

- **Bug Fixes**
	- Fixed an issue with the incorrect identification of invalid video URLs in the form validation.

- **Tests**
	- Updated existing tests to align with the new validation logic.
	- Added tests for transforming TUM-Live URLs into embeddable formats.

### CODE REVIEW

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/video-unit-form/video-unit-form.component.ts [hunk 1]
```diff
@@ -13,31 +13,35 @@ export interface VideoUnitFormData {
...
+function videoSourceTransformUrlValidator(control: AbstractControl): ValidationErrors | null {
+    const invalidVideoUrlError = { invalidVideoUrl: true };
+    const urlValue = control.value;
+    if (!urlValue) return null;
+    try {
+        return isTumLiveUrl(new URL(urlValue)) || urlParser.parse(urlValue) ? null : invalidVideoUrlError;
+    } catch {
+        return invalidVideoUrlError;
     }
+}
```
[the remark this answers is not in the document]
Comment (JohannesWt) [reply src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/video-unit-form/video-unit-form.component.ts:33]: The functions should return null if there is not error not the other way round. So the proposed implementation would not work

**Comment by JohannesWt** [comment]
The failing tests are not related to the changes I did.

**Review by nityanandaz — APPROVED** [review]
Manual test, locally.
Generates embeddable video URL correctly.

**Review by rstief — CHANGES_REQUESTED** [review]

**Review by basak-akan — APPROVED** [review]
Besides @rstief 's statements, code looks good 👍

**Review by milljoniaer — APPROVED** [review]
Discussed on slack, I am fine with this solution as well

**Review by basak-akan — APPROVED** [review]
Code lgtm 👍

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/video-unit-form/video-unit-form.component.ts [hunk 2]
```diff
@@ -13,31 +13,44 @@ export interface VideoUnitFormData {
...
+    const urlValue = control.value;
+    if (!urlValue) {
+        return undefined;
     }
-
+    let parsedUrl, url;
     try {
-        new URL(control.value);
+        url = new URL(urlValue);
+        parsedUrl = urlParser.parse(urlValue);
```
Comment (rstief) [inline src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/video-unit-form/video-unit-form.component.ts:32]: (Nitpick) again, this **could** be put after the try-catch block as it just returns undefined if the parsedUrl is not valid.

File: src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/video-unit-form/video-unit-form.component.ts [hunk 3]
```diff
@@ -21,27 +21,36 @@ function isVideoOnlyTumUrl(url: URL): boolean {
...
-        const urlValue = control.value;
-        const url = new URL(urlValue);
-        if (isTumLiveUrl(url)) return isVideoOnlyTumUrl(url) ? null : invalidVideoUrlError;
-        return urlParser.parse(urlValue) ? invalidVideoUrlError : null;
+        url = new URL(control.value);
     } catch {
-        return invalidVideoUrlError;
+        // intentionally empty
     }
+    if (url && !(isTumLiveUrl(url) && !isVideoOnlyTumUrl(url))) {
```
Comment (rstief) [inline src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/video-unit-form/video-unit-form.component.ts:50]: - we do not check the `urlParser.parse(urlValue)` anymore
- the if also correct if the URL is not from TumLive.

Are those things on purpose?

**Review by rstief — APPROVED** [review]
Code looks good after changes 👍

**Review by nityanandaz — APPROVED** [review]
Re-test after code changes, still works

**Review by MaximilianAnzinger — APPROVED** [review]
maintainer approved 👍

**Pull request merged**